pennylane/ops/qubit/time_dependent_hamiltonian.py

```diff
@@ -0,0 +1,218 @@
+"""
+This submodule contains the time-dependent hamiltonian class
+"""
+
+from jax import numpy as jnp
+
+import pennylane as qml
+from pennylane.operation import Observable, Tensor
+from pennylane.ops.qubit.hamiltonian import Hamiltonian
+
+
+# pylint: disable= too-many-instance-attributes
+class TDHamiltonian(Observable):
+    r"""Operator representing a time-dependent Hamiltonian.
+
+    The Hamiltonian is represented as a linear combination of other operators, e.g.,
+    :math:`\sum_{k=0}^{N-1} c_k O_k`, where the :math:`c_k` are trainable parameters,
+    some of which are time-dependent.
+
+    Args:
+        coeffs (Union[tensor_like, callable]): coefficients of the Hamiltonian expression.
+        observables (Iterable[Observable]): observables in the Hamiltonian expression, of same length as coeffs
+        id (str): Custom string to label a specific operator instance.
+        do_queue (bool): indicates whether the operator should be
+            recorded when created in a tape context
+
+    **Example:**
+
+    A time-dependent Hamiltonian can be created by passing a list of coefficients and callables,
+    as well as the list of observables.
+
+    >>> def f1(params, t): return jnp.polyval(params, t)
+    >>> coeffs = [0.2, f1]
+    >>> obs = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliY(0) @ qml.Identity(1)]
+    >>> H = TDHamiltonian(coeffs, obs)
+    >>> print(H)
+      (-0.543) [Z0 H2]   # ToDo: does not print nicely, and this output is incorrect
+    + (0.2) [X0 Z1]
+
+    The coefficients can be a trainable tensor, for example:  # ToDo: Can parameters be a trainable tensor? Update example regardless.
+
+    >>> coeffs = tf.Variable([0.2, -0.543], dtype=tf.double)
+    >>> obs = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.Hadamard(2)]
+    >>> H = qml.Hamiltonian(coeffs, obs)
+    >>> print(H)
+      (-0.543) [Z0 H2]
+    + (0.2) [X0 Z1]
+
+    In many cases, Hamiltonians can be constructed using Pythonic arithmetic operations.
+    For example:
+
+    >>> qml.Hamiltonian([1.], [qml.PauliX(0)]) + 2 * qml.PauliZ(0) @ qml.PauliZ(1)
+
+    is equivalent to the following Hamiltonian:
+
+    >>> qml.Hamiltonian([1, 2], [qml.PauliX(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+
+    Addition of TDHamiltonians with TDHamiltonians, Hamiltonians or other observables is possible with
+    tensor-valued coefficients, i.e.,
+
+    >>> H1 = qml.Hamiltonian(torch.tensor([1.]), [qml.PauliX(0)])
+    >>> H2 = qml.Hamiltonian(torch.tensor([2., 3.]), [qml.PauliY(0), qml.PauliX(1)])
+    >>> obs3 = [qml.PauliX(0), qml.PauliY(0), qml.PauliX(1)]
+    >>> H3 = qml.Hamiltonian(torch.tensor([1., 2., 3.]), obs3)
+    >>> H3.compare(H1 + H2)  # ToDo: compare function not implemented, modify documentation
+
+    # ToDo: do we want any grouping behaviour akin to options on Hamiltonian class?
```

Comment:
For the moment this is beyond the scope of the pulse project since we never compute expectation values of such a Parametrized Hamiltonian

Reply:
Great, thanks!